アルファベットaからzを順に0から25(10進法表記)に対応させ、さ
らにその各数値を5ビットの2進数に変換するとする。このとき、平文
「cat」を一文字ずつバーナム暗号の手順に従って暗号化するとする。
なお、暗号化鍵(2進法表記)は、1文字目は01010、2文字目は10000、
3文字目は00110とする。このとき、暗号文を2進数で答えよ。

In [2]:
def char_to_binary(char):
    # アルファベットを0から25に対応させる
    num = ord(char) - ord('a')
    # 5ビットの2進数に変換
    binary = format(num, '05b')
    return binary

def encrypt_char(char, key):
    # 文字を2進数に変換
    binary_char = char_to_binary(char)
    # 暗号化鍵とのXORを計算
    encrypted_char = ''.join(str(int(b) ^ int(k)) for b, k in zip(binary_char, key))
    return encrypted_char

def vernam_cipher(plaintext, keys):
    ciphertext = ""
    for char, key in zip(plaintext, keys):
        encrypted_char = encrypt_char(char, key)
        ciphertext += encrypted_char + ' '
    return ciphertext.strip()

# 平文 "cat" と暗号化鍵
plaintext = "cat"
keys = ["01010", "10000", "00110"]

# バーナム暗号で暗号化
ciphertext = vernam_cipher(plaintext, keys)

print("平文:", plaintext)
print("暗号文(binary):", ciphertext)


平文: cat
暗号文(binary): 01000 10000 10101


回答/Answer

cat ⇒ 2, 0, 19

⇒ 2進数(binary): 00010, 00000, 10011

各ビット列と暗号化鍵とのXORを計算 / Calculate XOR of
each bit string and encryption key
■00010⊕01010 = 01000

■00000⊕10000 = 10000

■10011⊕00110 = 10101

⇒暗号文(ciphertext): 01000 10000 10101